File: tier4_expert/greedy/problems/p478_generate_random_point_in_a_circle.py
"""
P478: Generate Random Point in a Circle [PREMIUM] (Medium)
https://leetcode.com/problems/generate-random-point-in-a-circle/
Topics: Math, Geometry, Rejection Sampling, Randomized

Given the radius and the position of the center of a circle, implement the function randPoint which generates a uniform random point inside the circle.
Implement the Solution class:
Example 1:
    Input
Example 1:
    ["Solution", "randPoint", "randPoint", "randPoint"]
    [[1.0, 0.0, 0.0], [], [], []]
    Output
    [null, [-0.02493, -0.38077], [0.82314, 0.38945], [0.36572, 0.17248]]

    Explanation
    Solution solution = new Solution(1.0, 0.0, 0.0);
    solution.randPoint(); // return [-0.02493, -0.38077]
    solution.randPoint(); // return [0.82314, 0.38945]
    solution.randPoint(); // return [0.36572, 0.17248]

Constraints:
    - 0 < radius <= 108
    - -107 <= x_center, y_center <= 107
    - At most 3 * 104 calls will be made to randPoint.

Template (python3):
    class Solution:

        def __init__(self, radius: float, x_center: float, y_center: float):


        def randPoint(self) -> List[float]:



    # Your Solution object will be instantiated and called as such:
    # obj = Solution(radius, x_center, y_center)
    # param_1 = obj.randPoint()
"""

import sys

sys.path.insert(0, ".")
from src.utils import Problem, TestCase


class Solution(Problem):
    name = "478. Generate Random Point in a Circle"
    test_cases = [
        TestCase(
            input={"radius": 1.0, "x_center": 0.0, "y_center": 0.0, "count": 1000},
            expected="in_circle",
            label="points within unit circle",
        ),
        TestCase(
            input={"radius": 0.5, "x_center": 1.0, "y_center": 2.0, "count": 1000},
            expected="in_circle",
            label="points within offset circle",
        ),
    ]

    def solve(self, radius: float, x_center: float, y_center: float, count: int) -> list[list[float]]:
        raise NotImplementedError("TODO: Implement solve(self, radius, x_center, y_center, count) -> list[list[float]]")


if __name__ == "__main__":
    Solution().run()

File: tier4_expert/greedy/problems/p478_generate_random_point_in_a_circle.rs
/*
 * P478: Generate Random Point in a Circle [PREMIUM] (Medium)
 * https://leetcode.com/problems/generate-random-point-in-a-circle/
 * Topics: Math, Geometry, Rejection Sampling, Randomized
 * 
 * Given the radius and the position of the center of a circle, implement the function randPoint which generates a uniform random point inside the circle.
 * Implement the Solution class:
 * Example 1:
 *     Input
 * Example 1:
 *     ["Solution", "randPoint", "randPoint", "randPoint"]
 *     [[1.0, 0.0, 0.0], [], [], []]
 *     Output
 *     [null, [-0.02493, -0.38077], [0.82314, 0.38945], [0.36572, 0.17248]]
 * 
 *     Explanation
 *     Solution solution = new Solution(1.0, 0.0, 0.0);
 *     solution.randPoint(); // return [-0.02493, -0.38077]
 *     solution.randPoint(); // return [0.82314, 0.38945]
 *     solution.randPoint(); // return [0.36572, 0.17248]
 * 
 * Constraints:
 *     - 0 < radius <= 108
 *     - -107 <= x_center, y_center <= 107
 *     - At most 3 * 104 calls will be made to randPoint.
 * 
 * Template (python3):
 *     class Solution:
 * 
 *         def __init__(self, radius: float, x_center: float, y_center: float):
 * 
 * 
 *         def randPoint(self) -> List[float]:
 * 
 * 
 * 
 *     # Your Solution object will be instantiated and called as such:
 *     # obj = Solution(radius, x_center, y_center)
 *     # param_1 = obj.randPoint()
 */
struct CircleSolver {
    radius: f64,
    x_center: f64,
    y_center: f64,
    state: u64,
}

impl CircleSolver {
    fn new(radius: f64, x_center: f64, y_center: f64) -> Self {
        let state = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_nanos() as u64)
            .unwrap_or(12345);
        Self { radius, x_center, y_center, state }
    }
    fn next_u64(&mut self) -> u64 {
        todo!()
    }
    fn rand_point(&mut self) -> (f64, f64) {
        todo!()
    }
}

fn main() {
    println!("\n============================================================");
    println!("  478. Generate Random Point in a Circle");
    println!("============================================================");
    let tests: &[(&str, f64, f64, f64, i32)] = &[
        ("points within unit circle", 1.0, 0.0, 0.0, 1000),
        ("points within offset circle", 0.5, 1.0, 2.0, 1000),
        ("tiny circle", 0.01, 0.0, 0.0, 100),
        ("large offset center", 5.0, -100.0, 200.0, 500),
        ("single point", 1.0, 0.0, 0.0, 1),
    ];
    let mut passed = 0;
    for (i, (label, r, cx, cy, count)) in tests.iter().enumerate() {
        let mut solver = CircleSolver::new(*r, *cx, *cy);
        let mut ok = true;
        for _ in 0..*count {
            let (px, py) = solver.rand_point();
            let dx = px - cx;
            let dy = py - cy;
            if dx * dx + dy * dy > r * r + 1e-9 { ok = false; break; }
        }
        if ok {
            passed += 1;
            println!("  Test {} ({}): PASS", i + 1, label);
        } else {
            println!("  Test {} ({}): FAIL", i + 1, label);
            println!("    Some points outside circle!");
        }
    }
    println!("\n  {}/{} passed", passed, tests.len());
    println!("============================================================\n");
    std::process::exit(if passed == tests.len() { 0 } else { 1 });
}
